tests/unit/test_charm.py

```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
 
     @patch("charm.apt.RepositoryMapping")
     @patch("charm.apt.DebianRepository.from_repo_line")
-    @patch("charm.urlopen")
-    def test_add_mongodb_org_repository_success(self, urlopen, from_repo_line, get_mapping):
+    @patch("charm.apt.DebianRepository.import_key")
+    def test_add_mongodb_org_repository_success(self, import_key, from_repo_line, repo_map):
+        repo_map.return_value = set()
+
         self.harness.charm._add_mongodb_org_repository()
-        get_mapping.return_value = {}
-        urlopen.assert_called()
         from_repo_line.assert_called()
+        (from_repo_line.return_value.import_key).assert_called_with(MONGODB_PUB_KEY)
+        self.assertEqual(repo_map.return_value, {from_repo_line.return_value})
 
     @patch("charm.apt.RepositoryMapping")
+    @patch("charm.apt.DebianRepository.from_repo_line")
+    @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.urlopen")
-    def test_add_mongodb_org_repository_gpg_fail_leads_to_blocked(self, urlopen, get_mapping):
-        get_mapping.return_value = {}
+    def test_add_mongodb_org_repository_gpg_fail_leads_to_blocked(
+        self, urlopen, import_key, from_repo_line, repo_map
+    ):
+        # preset values
+        repo_map.return_value = set()
         urlopen.side_effect = URLError("urlopen error")
-        with self.assertLogs("charm", "ERROR") as logs:
-            self.harness.charm._add_mongodb_org_repository()
-            self.assertIn("ERROR:charm:failed to get GPG key, reason:", "".join(logs.output))
+        self.harness.charm._add_mongodb_org_repository()
 
+        # verify we don't add repo when an exception occurs and that we enter blocked state
+        from_repo_line.assert_not_called()
```

Comment:
This and the next two lines are not really necessary and very much implementation dependent.  Just to illustrate, we could change the implementation so that the call to from_repo_line is made before urlopen, everything would still be correct but the unit test would fail.   I think here it is sufficient to check that the charm ended up in the BlockedStatus

Reply:
I agree that `from_repo_line` isn't necessary here!

But I think we should still check that we do not add the key, since if we get an error and add the key we're doing something wrong by adding to the repo when we shouldn't be. (The purpose of the function `_add_mongodb_org_repository` is to add a key to the repository mapping. If we get an error **we should NOT add the key** in any implementation of the code)

Similarly if we get a URL error from ` key = urlopen("...")` we will **not** want to call import key since that could cause problems later down the line, so making sure this doesn't get called here is also necessary. Since in any implementation we would not want to add a key that was the result of a `urlopen` error. (This should be true for all possible implementations of this function)